torch/csrc/jit/script/compiler.cpp

```diff
@@ -1333,6 +1331,54 @@ struct to_ir {
     }
   }
 
+  void emitIf(const If& stmt) {
+    // NOTE: emitIf checks on If stmt condition to see if the cond AST kind == is/is not,
+    // for such cases we do meta programming and disable emitting the corresponding branches
+    Expr cond = stmt.cond();
+
+    if (cond.kind() != TK_IS && cond.kind() != TK_ISNOT) {
+      // emit normal IF stmt for cases except TK_IS and TK_ISNOT
+      Value* cond_value = emitCond(cond);
+      emitIfElseBlocks(cond_value, stmt);
+    } else {
+      // meta programming on AST and emit branches base on the possible output of cond
+      auto cond_op = BinOp(cond);
+      SugaredValuePtr lhs_val = emitSugaredExpr(cond_op.lhs(), 1);
+      SugaredValuePtr rhs_val = emitSugaredExpr(cond_op.rhs(), 1);
+
+      List<Stmt> always_none_branch = cond.kind() == TK_IS? stmt.trueBranch(): stmt.falseBranch();
+      List<Stmt> never_none_branch = cond.kind() == TK_IS? stmt.falseBranch(): stmt.trueBranch();
+
+      if (lhs_val->isNone() && rhs_val->isNone()) {
+        // None is/is not None: only emit the always_none_branch
+        emitStatements(always_none_branch);
+      } else if (Value* v = asSimple(lhs_val)) {
+        if (v->type()->cast<OptionalType>()) {
```

Comment:
The check for OptionalType should be in the `SugaredValue::isNone` method. This weirdly spreads the responsibility across the compiler and sugared value to determine if something is none. That responsibility should be entirely in the sugared value.